File: scripts/deploy-netlify-v3.mjs
import fs from "node:fs";
import path from "node:path";
import crypto from "node:crypto";
import http from "node:http";
import https from "node:https";

const HOST = "api.netlify.com";
const SITE_ID = "00de2ab0-d070-47e8-9162-8e3558545393";
const AUTH_TOKEN = "nfp_3eDoe5u3X9NYEyzKv4fwhGQ7zzrtJmjMcdd8";
const PROXY_HOST = "127.0.0.1";
const PROXY_PORT = 18080;

function sha1(buf) {
  return crypto.createHash("sha1").update(buf).digest("hex");
}

function apiCall(method, urlPath, bodyBuf, contentType, timeoutMs = 120000) {
  return new Promise((resolve, reject) => {
    const req = http.request({
      host: PROXY_HOST,
      port: PROXY_PORT,
      method: "CONNECT",
      path: HOST + ":443",
      agent: false,
    });
    req.on("connect", (res, socket) => {
      if (res.statusCode !== 200) {
        reject(new Error("Proxy CONNECT failed: " + res.statusCode));
        return;
      }
      const req2 = https.request({
        hostname: HOST,
        port: 443,
        method,
        path: urlPath,
        socket,
        agent: false,
        servername: HOST,
        timeout: timeoutMs,
        headers: {
          "Authorization": "Bearer " + AUTH_TOKEN,
          "Content-Type": contentType || "application/json",
          "Content-Length": bodyBuf ? bodyBuf.length : 0,
          "User-Agent": "TraeDeploy/1.0",
        },
      });
      let data = "";
      req2.on("response", (res2) => {
        res2.on("data", (chunk) => (data += chunk));
        res2.on("end", () => resolve({ status: res2.statusCode, body: data }));
      });
      req2.on("error", reject);
      if (bodyBuf) req2.write(bodyBuf);
      req2.end();
    });
    req.on("error", reject);
    req.end();
  });
}

async function main() {
  console.log("\n=== 构建文件清单 ===\n");
  const deployFiles = {};
  const fileContents = {};

  // 1. HTML pages (from .netlify/deploy)
  const deployDir = "/workspace/.netlify/deploy/v1/blobs/deploy";
  if (fs.existsSync(deployDir)) {
    for (const name of fs.readdirSync(deployDir)) {
      const full = path.join(deployDir, name);
      const raw = fs.readFileSync(full);
      let decodedName;
      try { decodedName = Buffer.from(name, "base64").toString("utf8"); } catch { decodedName = name; }
      let htmlContent;
      try {
        const blob = JSON.parse(raw.toString("utf8"));
        htmlContent = blob.html || raw.toString("utf8");
      } catch {
        htmlContent = raw.toString("utf8");
      }
      let deployName;
      if (decodedName === "/index") deployName = "index.html";
      else if (decodedName === "/404.html" || decodedName === "404") deployName = "404.html";
      else if (decodedName === "/500.html" || decodedName === "500") deployName = "500.html";
      else if (decodedName.startsWith("/")) deployName = decodedName.slice(1) + "/index.html";
      else deployName = decodedName + "/index.html";

      const contentBuf = Buffer.from(htmlContent, "utf8");
      deployFiles[deployName] = sha1(contentBuf);
      fileContents[deployName] = { content: contentBuf, mime: "text/html; charset=utf-8" };
    }
  }

  // 2. Static assets (from .netlify/static)
  const staticDir = "/workspace/.netlify/static/_next";
  if (fs.existsSync(staticDir)) {
    const staticFiles = listDir(staticDir, "_next/");
    for (const name of staticFiles) {
      const full = path.join("/workspace/.netlify/static", name);
      const content = fs.readFileSync(full);
      const hash = sha1(content);
      const ext = path.extname(name).toLowerCase();
      const mime =
        ext === ".js" ? "application/javascript; charset=utf-8" :
        ext === ".css" ? "text/css; charset=utf-8" :
        ext === ".json" ? "application/json; charset=utf-8" :
        ext === ".png" ? "image/png" :
        ext === ".jpg" || ext === ".jpeg" ? "image/jpeg" :
        ext === ".svg" ? "image/svg+xml" :
        ext === ".woff" ? "font/woff" :
        ext === ".woff2" ? "font/woff2" :
        "application/octet-stream";
      deployFiles[name] = hash;
      fileContents[name] = { content, mime };
    }
  }

  // 3. _redirects file
  const redirects =
    "/api/* /.netlify/functions/___netlify-server-handler 200\n" +
    "/_next/data/* /.netlify/functions/___netlify-server-handler 200\n" +
    "/bots/* /.netlify/functions/___netlify-server-handler 200\n" +
    "/dashboard /.netlify/functions/___netlify-server-handler 200\n" +
    "/conversations/* /.netlify/functions/___netlify-server-handler 200\n" +
    "/settings /.netlify/functions/___netlify-server-handler 200\n" +
    "/analytics /.netlify/functions/___netlify-server-handler 200\n" +
    "/widget.js /.netlify/functions/___netlify-server-handler 200\n";
  const redirectsBuf = Buffer.from(redirects, "utf8");
  deployFiles["_redirects"] = sha1(redirectsBuf);
  fileContents["_redirects"] = { content: redirectsBuf, mime: "text/plain" };

  // 4. Function zip (26MB - we handle separately)
  const funcPath = "___netlify-server-handler/___netlify-server-handler.zip";
  const funcFull = "/workspace/.netlify/functions/___netlify-server-handler.zip";
  if (fs.existsSync(funcFull)) {
    const funcBuf = fs.readFileSync(funcFull);
    const funcHash = sha1(funcBuf);
    deployFiles[funcPath] = funcHash;
    fileContents[funcPath] = { content: funcBuf, mime: "application/zip" };
    console.log(`Function zip: ${funcBuf.length} bytes (${(funcBuf.length / 1024 / 1024).toFixed(1)}MB)`);
  }

  const total = Object.keys(fileContents).length;
  console.log(`\n共 ${total} 个文件待上传`);
  console.log(`  - HTML: ${Object.keys(deployFiles).filter(f => f.endsWith(".html")).length}`);
  console.log(`  - 静态资源: ${Object.keys(deployFiles).filter(f => f.startsWith("_next/")).length}`);
  console.log(`  - 重定向: 1`);
  console.log(`  - 函数: 1 (26MB)`);

  // Create deploy
  console.log("\n=== 创建 Deploy ===\n");
  const payload = Buffer.from(JSON.stringify({ files: deployFiles }), "utf8");
  const createResult = await apiCall(
    "POST",
    "/api/v1/sites/" + SITE_ID + "/deploys",
    payload,
    "application/json",
    120000
  );

  if (createResult.status < 200 || createResult.status >= 300) {
    console.error("错误:", createResult.status, createResult.body.slice(0, 500));
    process.exit(1);
  }
  const deployData = JSON.parse(createResult.body);
  const deployId = deployData.id;
  console.log("Deploy ID:", deployId);
  console.log("需上传:", deployData.required ? deployData.required.length : total, "个文件");

  console.log("\n=== 上传文件 ===\n");
  let idx = 0;
  for (const [fpath, info] of Object.entries(fileContents)) {
    idx++;
    const isFunc = fpath.startsWith("___netlify");
    const short = fpath.length > 60 ? fpath.slice(0, 60) + "..." : fpath;
    process.stdout.write(`  [${idx}/${total}] ${short} (${(info.content.length / 1024).toFixed(1)}KB)... `);

    const encodedPath = encodeURIComponent(fpath).replace(/%2F/g, "/");
    const timeout = isFunc ? 600000 : 120000;
    const result = await apiCall(
      "PUT",
      "/api/v1/deploys/" + deployId + "/files/" + encodedPath,
      info.content,
      info.mime,
      timeout
    );

    if (result.status >= 200 && result.status < 300) {
      console.log("OK");
    } else {
      console.log("FAILED (" + result.status + ")");
      if (result.body && result.body.length < 500) console.log("    " + result.body);
    }
  }

  console.log("\n=== 等待部署完成 ===\n");
  for (let tries = 0; tries < 60; tries++) {
    await new Promise((r) => setTimeout(r, 3000));
    const check = await apiCall("GET", "/api/v1/deploys/" + deployId, null, "application/json", 30000);
    try {
      const info = JSON.parse(check.body);
      console.log("  状态: " + info.state + " (locked: " + info.locked + ")");
      if (info.state === "ready" || info.state === "published") break;
    } catch {}
  }

  console.log("\n=== 完成 ===\n");
  console.log("Dashboard: https://app.netlify.com/sites/chatai-pro-2026/deploys/" + deployId);
  console.log("访问地址: https://chatai-pro-2026.netlify.app");
}

function listDir(dir, prefix) {
  const out = [];
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      out.push(...listDir(full, prefix + entry.name + "/"));
    } else {
      out.push(prefix + entry.name);
    }
  }
  return out;
}

main().catch((e) => { console.error(e); process.exit(1); });

File: scripts/deploy-netlify-v3.py
#!/usr/bin/env python3
import os, sys, json, hashlib, subprocess, urllib.parse, time

TOKEN = "nfp_3eDoe5u3X9NYEyzKv4fwhGQ7zzrtJmjMcdd8"
SITE_ID = "00de2ab0-d070-47e8-9162-8e3558545393"
PROXY = "http://127.0.0.1:18080"
API_BASE = "https://api.netlify.com"

def api_call(method, path, body_file=None, content_type="application/json", timeout=120):
    cmd = ["curl", "-sS", "-m", str(timeout),
           "-x", PROXY,
           "-H", f"Authorization: Bearer {TOKEN}",
           "-H", f"Content-Type: {content_type}",
           "-X", method,
           "-w", "\n%{http_code}",
           f"{API_BASE}{path}"]
    if body_file:
        cmd.extend(["--data-binary", f"@{body_file}"])

    result = subprocess.run(cmd, capture_output=True, check=False)
    raw = result.stdout.decode("utf-8", errors="replace")
    if "\n" in raw:
        body_part, status_part = raw.rsplit("\n", 1)
        try: status = int(status_part.strip())
        except ValueError:
            status = 0
            body_part = raw
    else:
        status = 0
        body_part = raw
    return status, body_part

def sha1_file(path):
    h = hashlib.sha1()
    with open(path, "rb") as f:
        while True:
            chunk = f.read(65536)
            if not chunk: break
            h.update(chunk)
    return h.hexdigest()

def sha1_bytes(data):
    return hashlib.sha1(data).hexdigest()

print("=== 构建文件清单 ===")
files = {}
file_paths = {}

# 1. 读取 HTML
deploy_dir = "/workspace/.netlify/deploy/v1/blobs/deploy"
if os.path.exists(deploy_dir):
    for name in os.listdir(deploy_dir):
        full = os.path.join(deploy_dir, name)
        raw = open(full, "rb").read()
        try:
            blob = json.loads(raw.decode("utf8"))
            html = blob["html"] if isinstance(blob, dict) and "html" in blob else raw.decode("utf8")
            html_bytes = html.encode("utf8")
        except:
            try: html_bytes = raw.decode("utf8").encode("utf8")
            except: html_bytes = raw

        try: fname = __import__("base64").b64decode(name).decode("utf8")
        except: fname = name

        if fname == "/index": deploy_name = "index.html"
        elif fname in ("/404.html", "404"): deploy_name = "404.html"
        elif fname in ("/500.html", "500"): deploy_name = "500.html"
        elif fname.startswith("/"): deploy_name = fname[1:] + "/index.html"
        else: deploy_name = fname + "/index.html"

        files[deploy_name] = sha1_bytes(html_bytes)
        tmp = f"/tmp/deploy_{abs(hash(deploy_name))}.html"
        with open(tmp, "wb") as f: f.write(html_bytes)
        file_paths[deploy_name] = (tmp, "text/html; charset=utf-8")

# 2. 读取静态资源
static_dir = "/workspace/.netlify/static/_next/static"
if os.path.exists(static_dir):
    for root, dirs, fnames in os.walk(static_dir):
        for fn in fnames:
            full = os.path.join(root, fn)
            rel = os.path.relpath(full, "/workspace/.netlify/static")
            files[rel] = sha1_file(full)
            ext = os.path.splitext(fn)[1].lower()
            mime = "application/octet-stream"
            if ext == ".js": mime = "application/javascript; charset=utf-8"
            elif ext == ".css": mime = "text/css; charset=utf-8"
            elif ext == ".json": mime = "application/json; charset=utf-8"
            elif ext == ".png": mime = "image/png"
            elif ext == ".svg": mime = "image/svg+xml"
            file_paths[rel] = (full, mime)

# 3. _redirects
redirects = (
    "/api/* /.netlify/functions/___netlify-server-handler 200\n"
    "/_next/data/* /.netlify/functions/___netlify-server-handler 200\n"
    "/bots/* /.netlify/functions/___netlify-server-handler 200\n"
    "/dashboard /.netlify/functions/___netlify-server-handler 200\n"
    "/conversations/* /.netlify/functions/___netlify-server-handler 200\n"
    "/settings /.netlify/functions/___netlify-server-handler 200\n"
    "/analytics /.netlify/functions/___netlify-server-handler 200\n"
    "/* /.netlify/functions/___netlify-server-handler 200\n"
).encode("utf-8")
redirects_file = "/tmp/_redirects"
with open(redirects_file, "wb") as f: f.write(redirects)
files["_redirects"] = sha1_bytes(redirects)
file_paths["_redirects"] = (redirects_file, "text/plain")

# 4. 函数 - 放在 files 里: netlify/functions/___netlify-server-handler.zip
func_file = "/workspace/.netlify/functions/___netlify-server-handler.zip"
func_sha = sha1_file(func_file)
func_size = os.path.getsize(func_file)
func_path = "netlify/functions/___netlify-server-handler.zip"
files[func_path] = func_sha
file_paths[func_path] = (func_file, "application/zip")
print(f"  {len(files)} 个文件 (含函数: {func_size} bytes = {func_size/1024/1024:.1f}MB)")

# 5. 创建 deploy
print("\n=== 创建 Deploy ===")
payload = json.dumps({
    "files": files,
    "functions": {
        "___netlify-server-handler": {
            "archive": "___netlify-server-handler.zip",
            "runtime": "nodejs24.x"
        }
    }
})
payload_file = "/tmp/deploy_payload.json"
with open(payload_file, "w") as f: f.write(payload)

status, body = api_call("POST", f"/api/v1/sites/{SITE_ID}/deploys", payload_file, "application/json", 120)
if status < 200 or status >= 300:
    print(f"错误 ({status}): {body[:500]}")
    sys.exit(1)

deploy = json.loads(body)
deploy_id = deploy["id"]
print(f"Deploy ID: {deploy_id}")
print(f"State: {deploy.get('state')}")
print(f"Required: {len(deploy.get('required', []))} files")
print(f"Required functions: {deploy.get('required_functions', [])}")

# 6. 上传所有文件 (包括函数)
print(f"\n=== 上传 {len(file_paths)} 个文件 ===")
failed = []
for i, (fpath, (filepath, mime)) in enumerate(file_paths.items()):
    is_func = "functions" in fpath
    timeout = 600 if is_func else 60
    encoded = urllib.parse.quote(fpath, safe="")
    status, body = api_call("PUT", f"/api/v1/deploys/{deploy_id}/files/{encoded}", filepath, mime, timeout)
    if 200 <= status < 300:
        if is_func:
            print(f"  [{i+1}/{len(file_paths)}] OK (FUNCTION) {fpath}")
        else:
            print(f"  [{i+1}/{len(file_paths)}] OK {fpath}")
    else:
        print(f"  [{i+1}/{len(file_paths)}] FAIL {fpath} ({status}): {body[:300]}")
        failed.append(fpath)

if failed:
    print(f"\n  重试 {len(failed)} 个失败文件...")
    for fpath in failed:
        filepath, mime = file_paths[fpath]
        is_func = "functions" in fpath
        timeout = 600 if is_func else 120
        encoded = urllib.parse.quote(fpath, safe="")
        status, body = api_call("PUT", f"/api/v1/deploys/{deploy_id}/files/{encoded}", filepath, mime, timeout)
        if 200 <= status < 300:
            print(f"    [重试] OK {fpath}")
        else:
            print(f"    [重试] FAIL {fpath} ({status}): {body[:300]}")

# 7. 等待部署完成
print("\n=== 等待部署完成 (最多 5 分钟) ===")
for i in range(50):
    time.sleep(6)
    status, body = api_call("GET", f"/api/v1/deploys/{deploy_id}", None, "application/json", 30)
    try:
        info = json.loads(body)
        state = info.get("state")
        print(f"  [{i+1}] 状态: {state}")
        if state in ("ready", "published"):
            break
        if state == "error":
            print(f"  部署错误! {info.get('error_message', '')}")
            break
    except:
        pass

print(f"\n=== 完成! ===")
print(f"站点: https://chatai-pro-2026.netlify.app")
print(f"Deploy: https://app.netlify.com/sites/chatai-pro-2026/deploys/{deploy_id}")
